client/network-gossip/src/bridge.rs

```diff
@@ -0,0 +1,299 @@
+// Copyright 2019 Parity Technologies (UK) Ltd.
+// This file is part of Substrate.
+
+// Substrate is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Substrate is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Substrate.  If not, see <http://www.gnu.org/licenses/>.
+
+use crate::Network;
+use crate::state_machine::{ConsensusGossip, Validator, TopicNotification};
+
+use network::Context;
+use network::message::generic::ConsensusMessage;
+use network::{Event, config::Roles, ReputationChange};
+
+use futures::{prelude::*, channel::mpsc, compat::Compat01As03, task::SpawnExt as _};
+use libp2p::PeerId;
+use parking_lot::Mutex;
+use sp_runtime::{traits::{Block as BlockT, NumberFor}, ConsensusEngineId};
+use std::{sync::Arc, time::Duration};
+
+pub struct GossipEngine<B: BlockT> {
+	inner: Arc<Mutex<GossipEngineInner<B>>>,
+	engine_id: ConsensusEngineId,
+}
+
+struct GossipEngineInner<B: BlockT> {
+	state_machine: ConsensusGossip<B>,
+	context: Box<dyn Context<B> + Send>,
+	context_ext: Box<dyn ContextExt<B> + Send>,
+}
+
+impl<B: BlockT> GossipEngine<B> {
+	/// Create a new instance.
+	pub fn new<N: Network<B> + Send + Clone + 'static>(
+		network: N,
+		executor: &impl futures::task::Spawn,
+		engine_id: ConsensusEngineId,
+		validator: Arc<dyn Validator<B>>,
+	) -> Self where B: 'static {
+		let mut state_machine = ConsensusGossip::new();
+		let mut context = Box::new(ContextOverService {
+			network: network.clone(),
+		});
+		let context_ext = Box::new(ContextOverService {
+			network: network.clone(),
+		});
+
+		// We grab the event stream before registering the notifications protocol, otherwise we
+		// might miss events.
+		let event_stream = network.event_stream();
+
+		network.register_notifications_protocol(engine_id);
+		state_machine.register_validator(&mut *context, engine_id, validator);
+
+		let inner = Arc::new(Mutex::new(GossipEngineInner {
+			state_machine,
+			context,
+			context_ext,
+		}));
+
+		let gossip_engine = GossipEngine {
+			inner: inner.clone(),
+			engine_id,
+		};
+
+		let res = executor.spawn({
+			let inner = Arc::downgrade(&inner);
+			async move {
+				loop {
+					let _ = futures_timer::Delay::new(Duration::from_millis(1100)).await;
+					if let Some(inner) = inner.upgrade() {
+						let mut inner = inner.lock();
+						let inner = &mut *inner;
+						inner.state_machine.tick(&mut *inner.context);
+					} else {
+						break;
```

Comment:
Can you add a note here that this branch is reached whenever the `GossipEngine` and thus the `GossipEngineInner` is dropped and thus the tick is not needed no more?